Signal/src/util/OWSWindowManager.m

```diff
@@ -0,0 +1,492 @@
+    OWSSingletonAssert();
+
+    return self;
+}
+
+- (void)setupWithRootWindow:(UIWindow *)rootWindow screenBlockingWindow:(UIWindow *)screenBlockingWindow
+{
+    OWSAssertIsOnMainThread();
+    OWSAssert(rootWindow);
+    OWSAssert(!self.rootWindow);
+    OWSAssert(screenBlockingWindow);
+    OWSAssert(!self.screenBlockingWindow);
+
+    self.rootWindow = rootWindow;
+    self.screenBlockingWindow = screenBlockingWindow;
+
+    self.returnToCallWindow = [self createReturnToCallWindow:rootWindow];
+    self.callViewWindow = [self createCallViewWindow:rootWindow];
+
+    [self ensureWindowState];
+}
+
+- (UIWindow *)createReturnToCallWindow:(UIWindow *)rootWindow
+{
+    OWSAssertIsOnMainThread();
+    OWSAssert(rootWindow);
+
+    // "Return to call" should remain at the top of the screen.
+    //
+    // TODO: Extend below the status bar.
+    CGRect windowFrame = rootWindow.bounds;
+    windowFrame.size.height = kReturnToCallWindowHeight;
+    UIWindow *window = [[UIWindow alloc] initWithFrame:windowFrame];
+    window.hidden = YES;
+    window.windowLevel = UIWindowLevel_ReturnToCall();
+    window.opaque = YES;
+    // This is the color of the iOS "return to call" banner.
+    // TODO: What's the right color to use here?
+    UIColor *backgroundColor = [UIColor colorWithRGBHex:0x4cd964];
+    window.backgroundColor = backgroundColor;
+
+    UIViewController *viewController = [OWSWindowRootViewController new];
+    viewController.view.backgroundColor = backgroundColor;
+
+    UIView *rootView = viewController.view;
+    rootView.userInteractionEnabled = YES;
+    [rootView addGestureRecognizer:[[UITapGestureRecognizer alloc] initWithTarget:self
+                                                                           action:@selector(returnToCallWasTapped:)]];
+
+    UILabel *label = [UILabel new];
+    label.text = NSLocalizedString(@"CALL_WINDOW_RETURN_TO_CALL", @"Label for the 'return to call' banner.");
+    label.textColor = [UIColor whiteColor];
+    // TODO: Dynamic type?
+    label.font = [UIFont ows_regularFontWithSize:14.f];
+    [rootView addSubview:label];
+    [label autoCenterInSuperview];
+    self.returnToCallLabel = label;
+
+    window.rootViewController = viewController;
+
+    return window;
+}
+
+- (UIWindow *)createCallViewWindow:(UIWindow *)rootWindow
+{
+    OWSAssertIsOnMainThread();
+    OWSAssert(rootWindow);
+
+    UIWindow *window = [[UIWindow alloc] initWithFrame:rootWindow.bounds];
+    window.hidden = YES;
+    window.windowLevel = UIWindowLevel_CallView();
+    window.opaque = YES;
+    // TODO: What's the right color to use here?
+    window.backgroundColor = [UIColor ows_materialBlueColor];
+
+    UIViewController *viewController = [OWSWindowRootViewController new];
+    viewController.view.backgroundColor = [UIColor ows_materialBlueColor];
+
+    UINavigationController *navigationController =
+        [[UINavigationController alloc] initWithRootViewController:viewController];
```

Comment:
Yeah, it's an interesting question.  I thought it would simplify the presentation of the call view since the navigation view controller would have reliable responsibility for making sure the call view had the right frame.  I don't want to bake in the assumption that we can switch a window's root view and that it will immediately have the correct frame.

Reply:
Seems reasonable, thanks for explaining.